Frontend only H5P Rendering and xAPI progress tracking integration

rtibbles opened the pull request:
## Summary

# H5P
* Currently, H5P rendering is executed as a Python reflection of https://github.com/tunapanda/h5p-standalone/
* This has the distinct advantage of working, and also reflecting the basic mechanism of how H5P is served on PHP
* However, because of design choices in how H5P is constructed, it tends to rely on hundreds of files to render a single content item, leading to an enormous number of simultaneous requests to the backend.
* It's not clear how well this will scale going forward, and is also not how the H5P PHP backend handles things
* The H5P PHP backend, which is the mostly widely used and well supported (https://github.com/h5p/h5p-php-library) relies on some PHP precaching of H5P assets - creating concatenated Javascript and CSS files that get served for each H5P content item. This allows a large reduction in the number of static assets served.
* We have recently implemented an entirely client side rendering flow for our own internal .perseus files (https://github.com/learningequality/kolibri-exercise-perseus-plugin/blob/release-v1.3.x/kolibri_exercise_perseus_plugin/assets/src/views/PerseusRendererIndex.vue#L423) - as our .perseus files and .h5p files are both just differently named zip files, a similar approach is taken here.
* Instead of using the zipcontent endpoint to load individual assets, the entire H5P file is preloaded to the frontend, and then unzipped, parsed, and each file turned into a Blob and a URL for that blob generated. (this is the same mechanism that [EpubJS](https://github.com/futurepress/epub.js/) uses to render Epubs in the frontend).
* We mirror the path substitution done in the PHP implementation for CSS concatenation, to ensure that all relatively referenced assets in CSS files are instead referenced by their new Blob URLs.
* For the user defined contents referenced in the H5P `content/content.json` we shim the `H5P.getPath` method to do a lookup into our own internal directory so we can return the Blob URLs.
* We also shim the H5P ContentType class to override the `getLibraryFilePath` which allows us to return our blob URLs for files inside the content widget library folders.

# xAPI
* H5P uses xAPI to log progress information and updates. In order to have a full integration for H5P, this PR also implements an xAPI statement validation layer, and a Javascript only API for xAPI.
* This does open up the possibility for HTML5 apps in general to use xAPI, but avoids the full blown Learning Record Store implementation that would be required to support the deprecated [xAPI Launch spec,](http://adlnet.github.io/xapi-launch/) and the newer [CMI5 Launch specification](https://github.com/AICC/CMI-5_Spec_Current/blob/quartz/cmi5_spec.md#content_launch)
* In line with our SCORM implementation, this adds validation and tests for that validation for xAPI statements
* It then adds a simple xAPI object accessible inside the inner iframe window, with an interface inspired by [xAPIWrapper](https://adlnet.github.io/xAPIWrapper/) but using Promises, and terminating in a local object store, while relying on general purpose hashi mechanisms to propagate it into saved state
* It then applies heuristics based on the [CMI5 vocabulary](https://github.com/AICC/CMI-5_Spec_Current/blob/quartz/cmi5_spec.md#xapi_data_model) and the verbs that are defined for use in H5P, in order to calculate progress
* It then enhances how Hashi reports progress to use either the xAPI calculated progress or the SCORM calculated progress

# zipcontent
* Serves all Hashi and H5P from static HTML files to avoid the complexity of creating paths for them on the WSGI server
* Hosts `zipcontent` and `static` files under the `content` path, to allow static versions of the alt origin server to be served seamlessly

Net result:

H5P loads with a single network request - albeit a longer one for large files.
H5P now reports progress to Kolibri.

Outstanding questions:
* What should we use for activity URIs for Kolibri content nodes? A URI based on a domain we own, see below.
* What should we use to identify Kolibri users to xAPI? An email address - see below.
* Should we define the `platform` property of the xAPI `context` object? No, this will add unnecessary bloat

Possible things that are still broken:
* I have not done an exhaustive check of H5P content, but have checked that all of the ones in @radinamatic's QA channel do indeed work
* I may have been too rigorous, or not rigorous enough in some of my validation work on the xAPI statements

TODOs:
* [x] Use zipcontent endpoint for the H5P file for getLibraryPath
* [x] Resizing the iframe so that the H5P does not inner scroll
* [x] Update User identifier to use a kolibri specific fake email
* [x] Update H5P static file paths to handle cache busting when we upgrade
* [x] Don't use contentNameSpace for H5P content identifier - just use a dummy value
* [x] Dummy activity id for xAPI statements
* [x] Strip out data before emitting from iframe
* [x] ~Reconstitute when bringing back in~ Not bothering with this as H5P never reads this data back in.
* [x] Prevent resize from emitting same value twice
* [x] Code split to load xAPI and H5P code on demand to avoid adding ~200KB to the Hashi bundle on every load just for H5P and xAPI
* [x] Concatenate all JS and CSS to prevent execution order issues
* [x] Fix residual display bugs for Course Presentation and Chart types
* [x] Handle errors during H5P rendering (e.g. Find the Words type breaks on IE11 due to internal incompatibilities in the H5P renderer code) and report an error to flag for the user


## References
Fixes #7776
Fixes #8127 

## Reviewer guidance
Is the code sufficiently well commented for posterity?
Where could more direct quotation of specification documents be helpful, could any of the foregoing essay be useful in the code?
Does H5P work as expected?
Do H5P activities record progress as expected?

----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [x] If there are any front-end changes, before/after screenshots are included
- [x] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests


## PR process

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [x] If this is an important user-facing change, PR or related issue has a 'changelog' label

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

rtibbles committed:
Render H5P files entirely on the clientside.

rtibbles committed:
Ensure that xAPI data is properly propagated.

rtibbles committed:
Add xAPI validation and tests.

rtibbles committed:
Add now transmission to baseShim.

rtibbles committed:
Implement simple first pass xAPI interface for internal use.

rtibbles committed:
Add tests and code comments to H5P implementation.

rtibbles committed:
Integrate H5P with our internal xAPI interface.

rtibbles committed:
Do better error logging of invalid statements.
Properly strip leading / from matched paths.

rtibbles committed:
Allow xAPI validators to simply warn.
Fix msg generation for correctResponse validation.

rtibbles committed:
Add xAPI progress tracking.

rtibbles committed:
Handle getLibraryFilePath calls.

rtibbles committed:
Simplify hashi and h5p template serving by using static files.

rtibbles committed:
Update frontend tests for new data storage, and xAPI validation warnings.

jamalex commented:
>The only thing we can't intercept is if a library calls the top level H5P.getLibraryPath method, and then tries to manually construct the path. The only work around I can think for this is for us to parse H5P files at import and keep a library of these files as static/media files somewhere, and have getLibraryPath point to that endpoint.

Could we not use the zipcontent endpoint as a fallback in this case?

>What should we use for activity URIs for Kolibri content nodes?

My suggestion would be something like `https://content.<somedomainweown>/<content_id>`.

- We want to use `content_id` rather than `node_id` because when we're logging interactions we want those to be scoped to the "identity" of the content itself, regardless of where it lives, with things like `node_id` and `channel_id` perhaps being included in the "context" fields.
- I don't think there's a strict requirement that these URLs serve anything in particular (or even exist), but having them on a domain we control means we could later choose to display something (or add a redirect) as useful.

>What should we use to identify Kolibri users to xAPI?

I'd still suggest (as we discussed briefly) something like a `mailto:<user_id>@<somedomainwecontrol>` (or the hashed version of that). Similarly, we don't need to have these emails be functional out of the gate, but we could choose to do something with them later.

rtibbles committed:
Move substatement schema definition after all dependency definitions.

rtibbles committed:
Handle #tmp suffixes in filenames.

jredrejo commented:
Tested with normal h5p files and h5p videos and checked this PR fixes #7776

rtibbles committed:
Add fallback to zipcontent endpoint for library url.

rtibbles committed:
Clean up content URLs and user identifiers.

rtibbles committed:
Generate H5P static files to allow cache busting.
Use our own H5P build to reduce complexity.

rtibbles committed:
Use mutation observer to trigger height container checks and resize.

rtibbles committed:
Add mutation observer shim for tests.

rtibbles committed:
Use a dummy activity id for H5P.
Strip out xAPI data we can infer when logging from H5P.

rtibbles committed:
Track the last sent height for resizes as a safeguard.

rtibbles committed:
Separate hashi shim interfaces so they can be loaded independently of implementations.
Use code splitting so xAPI and H5P code bloat doesn't affect every hashi load.

MBKayro commented:
The latest update fix a lot of problems, tested with ricecooker and H5P work perfectly and fast

rtibbles committed:
Use modern dynamic imports style.

rtibbles committed:
Update documentation for HTML5 apps.

rtibbles committed:
Pin H5P code to tagged versions.

rtibbles committed:
Add additional code comments.

rtibbles committed:
Minify h5p html css and JS.

rtibbles committed:
Miscellaneous cleanup from self review.

rtibbles committed:
Put zipcontent endpoint and zipcontent static files under /content/
for easier static deployment.

jamalex reviewed (approved):
@rtibbles walked us through the riskier bits on a call, and we were unable to... trip him up, so this seems good to me! The proof will be in some thorough manual testing pudding.

MBKayro commented on kolibri/core/content/utils/paths.py:
```diff
@@ -277,7 +277,7 @@ def zip_content_path_prefix():
     if path_prefix != "/":
         path_prefix = "/" + path_prefix
-    return path_prefix
+    return get_content_url(path_prefix)
```
i have this error : 
Uncaught (in promise) Error: Ajax error for /content/storage/6/b/6b843f1b8e7e37300fc3bce3dfa81706.h5p : 404 Not Found

and

Validation failed for 'download_url' in  
Object { id: Getter & Setter, contentnode: Getter & Setter, priority: Getter & Setter, preset: Getter & Setter, supplementary: Getter & Setter, thumbnail: Getter & Setter, checksum: Getter & Setter, available: Getter & Setter, file_size: Getter & Setter, extension: Getter & Setter, … }
mixin.js:61

rtibbles replied:
I've not seen the first issue - and that's precisely where the H5P file should be, so that's a little odd.

The second is a residual prop on the renderer mixin that is no longer being passed in, nothing to do with this work.

MBKayro replied:
this is what i see:

in a fresh installation, i have the first error to not load H5P in front-end, however the file missing is actually in content.

However, when i use      "return path_prefix" i can load h5p normaly, but when i switch to new "return get_content_url(path_prefix)" the H5P load and no error popup. so maybe it need a fresh content (.kolibri) to see error or maybe something i do wrong

rtibbles replied:
Thanks @MBKayro - the change I made was too aggressive for the result I was looking for, have updated it.

rtibbles committed:
Restore compat command.

rtibbles committed:
Polyfill URL.

rtibbles committed:
Polyfill Set.

MBKayro commented:
a lot of progress, and happy for the speed to load h5p, still to test in android device. Thanks @rtibbles for these great work

MBKayro commented:
it ill be great to send the "Total Score" of H5P activity to coach reports :)

rtibbles committed:
Ensure that deprecated libraries do not break due to changes in jquery API.

rtibbles committed:
Revert to pinning to commit for H5P vendoring, as releases are very out of date.
Include confirmation dialog and resizer.

rtibbles committed:
Handle iframe resizing in the html5 app viewer component.

rtibbles committed:
Add performance profiling debug log to H5P rendering.

rtibbles committed:
Use fflate instead of jszip.
Reduces build size by 100kb.
Increases H5P processing performance by ~20%

rtibbles committed:
Polyfill startsWith.

radinamatic commented:
H5P looking good on all browsers! 💯 

Two detailed noticed: warning about a deprecated H5P library, and a weird interaction where the content renderer section gets expanded as the user progresses through the activity and is pushing the rest of the page-container sections below, leaving empty space (happens in all the browsers).

IE11 | Chrome
---|---
![Win7 (start)  Running  - Oracle VM VirtualBox_059](https://user-images.githubusercontent.com/1457929/122839025-854b6d00-d2f7-11eb-9fa1-6228998afd55.jpg) | ![Win7 (start)  Running  - Oracle VM VirtualBox_061](https://user-images.githubusercontent.com/1457929/122839040-8da3a800-d2f7-11eb-84cd-6d1a8852e7df.jpg)

rtibbles commented:
The deprecation warning is expected.

The ever expanding content is not.

radinamatic commented:
> The deprecation warning is expected.

😎 

> The ever expanding content is not.

It's not _ever expanding_, just enough to be wasteful!

radinamatic commented:
Another comment that comes to mind: one of these resources took more than 15 seconds to load, but if full load is by design (if I understand correctly), is there a way for us to display that _loading_ status to the user?

rtibbles committed:
Only resize once after render for now.

rtibbles committed:
Add loading state for H5P apps.

radinamatic commented:
Yey for the loader, and much less wasted space!

Looks like IE11 is willing to load a video from YT, even though it's not available online, while Firefox & Chrome get blocked in the loading the player... 🤔 

IE11 | Firefox | Chrome
--- | --- | ---
![load-ie11](https://user-images.githubusercontent.com/1457929/122858271-4c70bf80-d31a-11eb-9409-cc547de937b8.gif) | ![Win7 (start)  Running  - Oracle VM VirtualBox_066](https://user-images.githubusercontent.com/1457929/122858309-5d213580-d31a-11eb-8316-abc038b0d922.jpg) | ![Win7 (start)  Running  - Oracle VM VirtualBox_067](https://user-images.githubusercontent.com/1457929/122858334-67dbca80-d31a-11eb-9e84-b158e1242d6a.jpg)

A part from this video hickup, let's wait for @pcenov to chime in about testing on a wider variety of H5P resources, before we merge!

pcenov commented:
@radinamatic I tested the rendering of the H5P resources in IE11 and Chrome on Windows 7 by importing the H5P Samples channel and did not observe any IE11 specific issues.
However I stumbled upon multiple issues that are valid for all browsers such as:
1. Some of the resources like Chart, Collage and Course Presentation do not load at all:
![2021-06-22_12-43-56](https://user-images.githubusercontent.com/79847249/122924195-2ecd4580-d36e-11eb-8631-b6f3fb9930db.png)

2. The first card of the Dialog Cards resource doesn't load until I click the Next arrow:
![2021-06-22_15-25-04](https://user-images.githubusercontent.com/79847249/122924663-ae5b1480-d36e-11eb-86f2-a4f8567ab8a3.png)

3. The Virtual Tour (360) doesn't seem to work in both browsers and in Chrome it doesn't even load:
![2021-06-22_15-34-10](https://user-images.githubusercontent.com/79847249/122925719-c67f6380-d36f-11eb-9bcb-22d42bd7fb5b.png)

4. I'm seeing various errors in the Console such as:

`TypeError: Cannot read property 'icon' of undefined`
`Error: Kolibri/core/assets/src/views/ContentRenderer/filePresetStrings.js] Download translation string not defined for 'h5p'`
`Uncaught TypeError: Cannot read property 'on' of undefined`
`Unable to find constructor for: H5P.Chart 1.2`

5. Occasional 400 error:

![2021-06-22_15-44-07](https://user-images.githubusercontent.com/79847249/122926984-1874b900-d371-11eb-9ec3-76700ec64dce.png)

rtibbles committed:
Enforce ordering of assets within a package.
Do CSS and JS asset concatenation to better mirror PHP implementation,
and prevent setTimeout unexpected behaviour.

rtibbles committed:
Prevent prematurely thinking that the H5P html has loaded.

rtibbles committed:
Handle abuse of getLibraryFilePath.

rtibbles committed:
Resize after loading has finished to force fit.

rtibbles committed:
Generate the zipcontent url once.
Shim deprecated getContentLibraryPath.

rtibbles committed:
Skip dependencies that don't exist.

pcenov commented:
@radina I am able to install [the latest build](https://buildkite.com/organizations/learningequality/pipelines/kolibri-windows/builds/1326/jobs/14f7d3b9-6010-4d59-87b3-323835175519/artifacts/5f4ba2ca-47b3-4a6a-a33c-501ce89839b2) on Windows 7, run it in IE11, import the 43 H5P resources from the H5P Samples channel and see most of them.
The following resources still won’t load: Chart, Collage, Course Presentation, Essay, Find Multiple Hotspots, Find the Words, Image Hotspots, Image Juxtaposition, Impressive Presentation, Single Choice Set
The Dialog Cards and the Virtual Tour (360) issues are still present.

MBKayro commented:
hi @pcenov  i have the same problem, however i think  that the channel must be updated with the latest H5P plugins. i generate new H5P content and upload it with recicooker and use it to test and most of H5P work perfectly like presentation ....

rtibbles commented:
In my local testing, I had previously seen issues with several of the H5P content types, but had managed to eliminate them.

I think there may be some timing/performance issues happening where in some machines things are happening in a different order, and this is causing things to present differently.

@MBKayro - yes, I imagine upgrading the content might resolve many issues, but I am also keen to make it as backwards compatible as possible, as, unlike an online instance, Kolibri will not be able to frequently update the H5P packages.

radinamatic commented:
H5P | Browser | Screenshot
--- | ---- | ---
Accordion | - FF ❌</br>   - Chr. ✔️ </br> - IE11 ✔️  | ![Win7 (start)  Running  - Oracle VM VirtualBox_086](https://user-images.githubusercontent.com/1457929/123647553-79e0d000-d828-11eb-90d9-6580a85e0a10.jpg)
Advanced Fill the Blanks | - FF ❌</br>   - Chr. ✔️ </br> - IE11 ✔️  | ![Win7 (start)  Running  - Oracle VM VirtualBox_087](https://user-images.githubusercontent.com/1457929/123648506-49e5fc80-d829-11eb-97cd-bbacf6aefe8b.jpg)
Arithmetic Quiz</br> </br> **note the differences between Chrome & IE11** | - FF ❌</br>   - Chr. ✔️ </br> - IE11 ✔️  | ![Win7 (start)  Running  - Oracle VM VirtualBox_088](https://user-images.githubusercontent.com/1457929/123649015-bd880980-d829-11eb-9f21-e68f9dea6178.jpg)
Chart | - FF ❌</br>   - Chr. ❌ </br> - IE11 ❌  | ![Win7 (start)  Running  - Oracle VM VirtualBox_089](https://user-images.githubusercontent.com/1457929/123650713-20c66b80-d82b-11eb-9858-73223a9b9261.jpg)
Collage | - FF ❌</br>   - Chr. ✔️ </br> - IE11 ❌  | ![Win7 (start)  Running  - Oracle VM VirtualBox_090](https://user-images.githubusercontent.com/1457929/123653837-cb3f8e00-d82d-11eb-9cd4-3d5ca5eff527.jpg)
Course Presentation | - FF ✔️</br>   - Chr. ✔️ </br> - IE11 ❌  | ![Win7 (start)  Running  - Oracle VM VirtualBox_091](https://user-images.githubusercontent.com/1457929/123654957-dd6dfc00-d82e-11eb-8913-74e232344d3a.jpg)

I'll continue, mixed bag so far... 😕

rtibbles committed:
Move all H5P config into H5P bundle to prevent overwriting.
Fixes loading bug on Firefox.

rtibbles committed:
Code quality cleanup.

radinamatic commented:
With the latest asset I found one activity (*Find the Words*) that fails in Firefox, and another (*Interactive Video*) that fails to load in both FF & Chrome on Windows 7.

FF | FF | Chrome
--- | --- | ---
![Win7 (start)  Running  - Oracle VM VirtualBox_002](https://user-images.githubusercontent.com/1457929/125328353-8430bd00-e344-11eb-92c0-2102d4971b8c.png) | ![Win7 (start)  Running  - Oracle VM VirtualBox_003](https://user-images.githubusercontent.com/1457929/125328391-8c88f800-e344-11eb-9196-434453be1152.png) | ![Win7 (start)  Running  - Oracle VM VirtualBox_004](https://user-images.githubusercontent.com/1457929/125328420-93b00600-e344-11eb-8542-4269a0ba98f4.png)

rtibbles committed:
Always show iframe to simplify rendering and prevent:
https://bugzilla.mozilla.org/show_bug.cgi?id=941146
Remove resize event triggering now that v-show is not being used.

rtibbles committed:
Simplify LOADING hashi state.
Add h5p error handling.

pcenov commented:
Hi @radinamatic 
Here’s what I found while testing the latest build on Windows 7 and Windows 10 https://buildkite.com/learningequality/kolibri-windows/builds/1379#5079953e-2c76-4117-a8cb-322562e1bb88 using Firefox, IE11, Edge and Chrome

1. The H5p resources are working significantly better than the previous time I checked. Here are the issues I was able to find:

WIN 7:

Firefox 90.0 (32-bit):

For some reason there are multiple H5P resources which are not displayed correctly in Firefox on Win 7:
Arithmetic Quiz, Chart, Drag the Words, Impressive Presentation, Personality Quiz (zoomed in too much), Single Choice Quiz
![2021-07-13_12-16-26](https://user-images.githubusercontent.com/79847249/125473651-94fd14d8-dc79-4e56-85e2-cdd8778b720a.png)
![2021-07-13_12-27-51](https://user-images.githubusercontent.com/79847249/125473656-a5283fd9-f899-45a4-bef1-7a972f48be15.png)
![2021-07-13_12-36-51](https://user-images.githubusercontent.com/79847249/125473658-ddb6e9a3-30b6-4a59-afbc-86bafab0efd4.png)
![2021-07-13_12-53-52](https://user-images.githubusercontent.com/79847249/125473661-a9f60798-f153-4534-b203-65be34c66eb5.png)
![2021-07-13_13-31-30](https://user-images.githubusercontent.com/79847249/125473682-f84caa8e-a078-4651-9463-40e30305b22d.png)
![2021-07-13_13-46-20](https://user-images.githubusercontent.com/79847249/125473686-06153699-e4a8-4af1-83e3-20e6ffe506b6.png)
![2021-07-13_14-07-52](https://user-images.githubusercontent.com/79847249/125473688-14ed9c8a-1ba1-4425-b962-22f432982b0e.png)

IE 11

Find the Words is stuck loading -- known issue, as discussed

The interactive Video resource is displayed only in IE11 --- known issue, as discussed

WIN 10

IE 11
The Impressive Presentation resource won’t load:
![2021-07-13_12-16-26](https://user-images.githubusercontent.com/79847249/125473848-14643967-d92e-47e0-9887-550269b7ef2a.png)

2. Loading of audio files in the Kamakalima channel - I did not experience any issues in any of the four browsers, the audio and resources were displayed correctly.

rtibbles commented:
> For some reason there are multiple H5P resources which are not displayed correctly in Firefox on Win 7:
Arithmetic Quiz, Chart, Drag the Words, Impressive Presentation, Personality Quiz (zoomed in too much), Single Choice Quiz

I think this must be to do with resizing logic. I will try reinstating the fix for this once more.

For the Timeline - the images not displaying is due to their being loaded from an external site, this is expected and a wontfix. IE11 is odd in that it lets the images through in spite of the content security policy.

> The Impressive Presentation resource won’t load:

My guess is that this is specific to IE11 and outside of our control. I verified this by attempting to render https://h5p.org/impressive-presentation (the source of the demo content) in BrowserStack IE11, it too failed to load, so this is a wontfix, as it's outside of our control.

rtibbles committed:
Reinstate post render resize trigger.

radinamatic commented:
With the last asset from yesterday I was experiencing similar issues in Firefox as @pcenov above, but it would be inconsistent upon page reload, resource appearing squished or zoomed in, then loading fine again, then squished again... 

The latest asset from today managed to fix that for all resources minus the **Personality Quiz**:

first load | reload 1 | reload 2 | reload 3 | reload 4
--- | --- | --- | --- | ---
![Win7 (start)  Running  - Oracle VM VirtualBox_022](https://user-images.githubusercontent.com/1457929/125511095-6c1c3c17-7814-4903-adce-d3ee590790de.png) | ![Win7 (start)  Running  - Oracle VM VirtualBox_023](https://user-images.githubusercontent.com/1457929/125511091-dcbc85e6-f4d0-4db7-9a30-d6b5eabf602f.png) | ![Win7 (start)  Running  - Oracle VM VirtualBox_024](https://user-images.githubusercontent.com/1457929/125511087-6d85f0b6-d59d-4bcc-be6c-03810d088766.png) | ![Win7 (start)  Running  - Oracle VM VirtualBox_026](https://user-images.githubusercontent.com/1457929/125510674-68ccbbaa-e5f8-4c42-a754-550926d68341.png) | ![Win7 (start)  Running  - Oracle VM VirtualBox_025](https://user-images.githubusercontent.com/1457929/125510675-e77de192-5705-4bcc-a0b0-e89b164d77cd.png)

A part from that, all the rest is good to go! 👏🏽

rtibbles commented:
I think I'm happy to leave "inconsistent scaling of background image on the Personality Quiz type" as a known issue...

radinamatic reviewed (approved):
> I think I'm happy to leave "inconsistent scaling of background image on the Personality Quiz type" as a known issue...

I wholeheartedly agree, approved! 🎉

rtibbles commented:
Failed MacOS Python test check appears to be a weird blip. Merging.
